tool/tsh/common/git.go

```diff
@@ -19,24 +19,119 @@
 package common
 
 import (
+	"bytes"
+	"io"
+	"os/exec"
+	"strings"
+
 	"github.com/alecthomas/kingpin/v2"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
 )
 
 type gitCommands struct {
-	list  *gitListCommand
-	login *gitLoginCommand
+	list   *gitListCommand
+	login  *gitLoginCommand
+	ssh    *gitSSHCommand
+	config *gitConfigCommand
+	clone  *gitCloneCommand
 }
 
 func newGitCommands(app *kingpin.Application) gitCommands {
 	git := app.Command("git", "Git server commands.")
 	cmds := gitCommands{
-		login: newGitLoginCommand(git),
-		list:  newGitListCommand(git),
+		login:  newGitLoginCommand(git),
+		list:   newGitListCommand(git),
+		ssh:    newGitSSHCommand(git),
+		config: newGitConfigCommand(git),
+		clone:  newGitCloneCommand(git),
 	}
 
 	// TODO(greedy52) hide the commands until all basic features are implemented.
 	git.Hidden()
 	cmds.login.Hidden()
 	cmds.list.Hidden()
+	cmds.config.Hidden()
+	cmds.clone.Hidden()
 	return cmds
 }
+
+type gitSSHURL struct {
+	*transport.Endpoint
+}
+
+func (g gitSSHURL) check() error {
+	switch {
+	case g.isGitHub():
+		if err := types.ValidateGitHubOrganizationName(g.owner()); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+func (g gitSSHURL) isGitHub() bool {
+	return g.Host == "github.com"
+}
+
+// owner returns the first part of the path. If the path does not have an owner,
+// an empty string is returned.
+//
+// For GitHub, owner is either the user or the organization that owns the repo.
+func (g gitSSHURL) owner() string {
+	owner, _, ok := strings.Cut(strings.TrimPrefix(g.Path, "/"), "/")
```

Comment:
How reliable is this? I'm worried about unintended consequences if `g.Path` is shaped differently from our expectations. Perhaps we can ensure some well-known prefix is present?